### PR TITLE
Add asset api integration

### PR DESCRIPTION
**Pull request by maysakanoni**

### CODE REVIEW

File: HousingManagementSystemApi/Startup.cs [hunk 1]
```diff
@@ -8,15 +8,20 @@
...
     using Gateways;
+    using Hackney.Shared.Asset.Domain;
     using HousingRepairsOnline.Authentication.DependencyInjection;
     using UseCases;
 
     public class Startup
     {
         private const string HousingManagementSystemApiIssuerId = "Housing Management System Api";
-
+        public static IEnumerable<AssetType> EligibleAssetTypes = new[]
```
Comment (maysakanoni) [inline HousingManagementSystemApi/Startup.cs:21]: if i make this configurable i think this will have to be a string and not IEnumerable

File: HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            "{\"id\":\"c7f39282-2505-f0fb-fbe3-d4ae3c8b7097\",\"assetId\":\"AssetId\",\"assetType\":\"Dwelling\",\"rootAsset\":\"ROOT\",\"parentAssetIds\":\"ROOT\",\"assetLocation\":{\"floorNo\":\"\",\"totalBlockFloors\":0,\"parentAssets\":[]},\"assetAddress\":{\"uprn\":\"100022574652\",\"addressLine1\":\"97A Old Church Road\",\"addressLine2\":\"Chingford\",\"addressLine3\":\"LONDON\",\"addressLine4\":\"\",\"postCode\":\"E4 6ST\",\"postPreamble\":\"\"},\"assetManagement\":{\"agent\":\"Homefinders\",\"areaOfficeName\":null,\"isCouncilProperty\":false,\"managingOrganisation\":\"London Borough of Hackney\",\"managingOrganisationId\":\"c01e3146-e630-c2cd-e709-18ef57bf3724\",\"owner\":\"\",\"isTMOManaged\":false,\"propertyOccupiedStatus\":\"VR\",\"isNoRepairsMaintenance\":false},\"assetCharacteristics\":{\"numberOfBedrooms\":3,\"numberOfLifts\":0,\"numberOfLivingRooms\":0,\"windowType\":\"\",\"yearConstructed\":\"0\"},\"tenure\":{\"id\":\"6af3ff4e-09de-f27d-514e-20de013b207a\",\"paymentReference\":\"9174652206\",\"type\":\"Temp Private Lt\",\"startOfTenureDate\":\"2013-06-24T00:00:00\",\"endOfTenureDate\":\"2015-02-01T00:00:00\",\"isActive\":false}}";
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs:21]: Please obfuscate identifying information to be random and/or generic data.

File: HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        public static IEnumerable<object[]> InvalidArgumentTestData()
+        {
+            yield return new object[] { new ArgumentNullException(), null };
+            yield return new object[] { new ArgumentException(), "" };
+            yield return new object[] { new ArgumentException(), " " };
+        }
+
+        [Fact]
+#pragma warning disable CA1707
+        public async void GivenAAssetIdArgument_WhenRetrievingAnAsset_ThenNoExceptionIsThrown()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs:65]: Please rename to be grammatically correct (A -> An), i.e.
```suggestion
        public async void GivenAnAssetIdArgument_WhenRetrievingAnAsset_ThenNoExceptionIsThrown()
```

File: HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            // Act
+            Func<Task> act = async () => await systemUnderTest.RetrieveAsset(AssetId);
+
+            // Assert
+            await act.Should().NotThrowAsync();
+        }
+
+        [Fact]
+#pragma warning disable CA1707
+        public async void GivenAAssetId_WhenRetrievingAnAsset_ThenAddressesAreRetrievedFromApi()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs:79]: Please rename to be grammatically correct (A -> An), i.e.
```suggestion
        public async void GivenAnAssetId_WhenRetrievingAnAsset_ThenAddressesAreRetrievedFromApi()
```

File: HousingManagementSystemApi.Tests/UseCasesTests/RetrieveAddressesUseCaseTests.cs [hunk 5]
```diff
@@ -32,15 +41,33 @@ public async Task GivenAPostcode_WhenExecute_GatewayReceivesCorrectInput()
...
+
+            retrieveAssetGateway.Setup(x => x.RetrieveAsset("assetId"))
+                .ReturnsAsync(new AssetResponseObject { AssetType = AssetType.Dwelling });
+
             var result = await retrieveAddressesUseCase.Execute(TestPostcode);
             result.First().PostalCode.Should().Be(TestPostcode);
         }
 
+        [Fact]
+        public async Task GivenAPostcode_WhenAnAddressExistsWithAIneligibleAssetType_ThenAPropertyIsNotReturned()
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/UseCasesTests/RetrieveAddressesUseCaseTests.cs:58]: Please rename to be grammatically correct (A -> An), i.e.
```suggestion
        public async Task GivenAPostcode_WhenAnAddressExistsWithAnIneligibleAssetType_ThenAPropertyIsNotReturned()
```

File: HousingManagementSystemApi/Constants.cs [hunk 6]
```diff
@@ -1,5 +1,9 @@
 namespace HousingManagementSystemApi;
 
+using System.Collections.Generic;
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Constants.cs:3]: Without any other code changes in this file, these new using statements seem redundant.
Please review and remove if they aren't necessary.

File: HousingManagementSystemApi/Gateways/AddressesHttpGateway.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
+using System.Collections.Generic;
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Gateways/AddressesHttpGateway.cs:1]: This class and file have been renamed (to `HousingSearchGateway`) and doesn't need to be reintroduced.

Please delete this file.

File: HousingManagementSystemApi/Startup.cs [hunk 8]
```diff
@@ -8,15 +8,20 @@
...
+        {
+            AssetType.Flat, AssetType.House, AssetType.Dwelling,
+        };
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Startup.cs:24]: Nitpick: Please add a new line afterwards to maintain readability

File: HousingManagementSystemApi/Startup.cs [hunk 9]
```diff
@@ -77,8 +91,9 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
         private static void AddHttpClients(IServiceCollection services)
         {
-            AddHttpClient(services, HttpClientNames.HousingSearch, "HOUSING_SEARCH_API_URI", "HOUSING_SEARCH_API_KEY");
             AddHttpClient(services, HttpClientNames.Properties, "PROPERTIES_API_URI", "PROPERTIES_API_KEY");
+            AddHttpClient(services, HttpClientNames.Asset, "HOUSING_ASSET_API_URI", "HOUSING_ASSET_API_KEY");
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Startup.cs:95]: Nitpick: Restore original ordering to remove unrelated changes and maintain atomic commit.

File: HousingManagementSystemApi/Startup.cs (hunk 1 above)
Comment (Bhavesh-Patel) [reply HousingManagementSystemApi/Startup.cs:21]: I believe you mean it would be a `string` and not `AssetType`.

**TL;DR It's fine the way it is.**

Configuration values can be the values of Enums.
If a single Enum value is value is required, then the value should be a string/text of the value.
However, if multiple Enums values are required to be specified in a single configuration value, the following would be required:
- The [FlagsAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=net-6.0) applied to the Enum
- The configuration value specified using comma separated values of the Enum values.
- [Optional, but recommended] Enum values be assigned unique, non-combinable integer values to avoid accidental combinations, i.e. 0, 1, 2, 4, 8 (easier to use [bitwise shifting](https://www.alanzucconi.com/2015/07/26/enum-flags-and-bitwise-operators/#:~:text=Ice%3B%20%20//%20Toggle%20ice-,Bitwise%20shifts,-The%20last%20two) to specify these values)

So addition info on this [StackOverflow answer](https://stackoverflow.com/a/13176549).

As this enum, `AssetType`, comes from an external library which making changes to would be a larger scope of work, this is fine the way it is.

File: HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs (hunk 2 above)
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi.Tests/GatewaysTests/AssetGatewayTests.cs:21]: This response contains a lot more information than is used and reducing it to just the require data would improve  readability.

**Review by Bhavesh-Patel — CHANGES_REQUESTED** [review]

File: HousingManagementSystemApi/Properties/launchSettings.json [hunk 10]
```diff
@@ -28,6 +28,8 @@
         "AUTHENTICATION_IDENTIFIER": "<REPLACE WITH A UNIQUE AUTHENTICATION IDENTIFIER>",
         "JWT_SECRET" : "<REPLACE WITH A SECRET>",
         "SENTRY_DNS": "https://<REPLACE WITH SENTRY DSN>",
+        "HOUSING_ASSET_API_URI": "https://<REPLACE WITH API URI>",
```
Comment (Bhavesh-Patel) [inline HousingManagementSystemApi/Properties/launchSettings.json:31]: Serverless.yml will need to have these environment variables assigned via AWS Parameter Store too.

**Review by Bhavesh-Patel — APPROVED** [review]
